### ISSUE TITLE
Internal error at src/full/Agda/TypeChecking/Conversion.hs:391

### ISSUE THREAD

**Issue by Saizan** [agda/agda#4994]
Error originally reported by Egbert Rijke on zulip https://agda.zulipchat.com/#narrow/stream/238741-general/topic/internal.20error

`computeElimHeadType` throws this error when called on a projection-like with no arguments provided.

The original error was about conversion checking I believe, but it was easier to produce a small test case with `--rewriting`.
 
```agda
{-# OPTIONS --rewriting #-}
module _ where

open import Agda.Builtin.Equality
open import Agda.Builtin.Equality.Rewrite
open import Agda.Builtin.Bool

data D (A : Set) : Set where
  con : (a : A) → D A

pr : ∀ {A} → D A → A
pr (con a) = a

postulate
  f : D Bool → Bool
  prf : pr ≡ f

{-# REWRITE prf #-}
```

```
An internal error has occurred. Please report this as a bug.
Location of the error: __IMPOSSIBLE__, called at src/full/Agda/TypeChecking/Conversion.hs:391:20 in Agda-2.6.2-LvRmYY9su5wDW05FKoAKoV:Agda.TypeChecking.Conversion
```

**Comment by jespercockx** [comment]
I guess we shouldn't call `computeElimHeadType` when both lists of eliminations are empty. There are two calls to this function, one of them is properly guarded but the other one (from `antiUnify`) is not. So I guess the easiest thing to do would be to add an extra case for `(Def f [] , Def g [])` to the definition of `antiUnify`.

**Comment by Saizan** [comment]
`computeElimHeadType` could be made total if it also took the type of the full application as an argument btw, I wasn't sure if that was available at both call sites though.

**Comment by jespercockx** [comment]
It is available at both call sites, so making it total would be even better.

**Comment by andreasabel** [comment]
Agda versions 2.5.2-2.6.1 report `prf is not a legal rewrite rule`.  (Agda-2.5.1.1 accepts.)

**Comment by jespercockx** [comment]
When was anti-unification added? I wouldn't be surprised if that coincides with the introduction of this bug.

**Comment by vlopezj** [comment]
For what it's worth, I have removed the antiunification code from my branch (https://github.com/agda/agda/tree/future-hetcon-no-compat ), and the error still triggers.   I think the error as reported by Andreas is now triggered from `Agda.TypeChecking.Rewriting.checkRewriteRule`.

**Comment by andreasabel** [comment]
> and the error still triggers. 

But not the internal error, or?

> I think the error as reported by Andreas is now triggered from `Agda.TypeChecking.Rewriting.checkRewriteRule`.

Well, if Agda complains about `prf is not a legal rewrite rule`, that would be ok.

Can the original internal error still be triggered?

> Error originally reported by Egbert Rijke on zulip [agda.zulipchat.com/#narrow/stream/238741-general/topic/internal.20error](https://agda.zulipchat.com/#narrow/stream/238741-general/topic/internal.20error)

**Comment by Saizan** [comment]
The error was triggered by this revision of the linked repo
https://github.com/HoTT-Intro/Agda/commit/24f6861188f1405df7c3f673b5192a3d17016254

I tried to compile the hetcon branch, but I got an error from GHC, so I couldn't test this myself:
```
rc/full/Agda/TypeChecking/Conversion.hs:2230:3: warning:
    Pattern match checker exceeded (2000000) iterations in
    an equation for ‘go’. (Use -fmax-pmcheck-iterations=n
    to set the maximum number of iterations to n)
     |
2230 |   go (SingleT (H'Both (Pi dom cod))) = return$ TPi (asTwin dom) (asTwin cod)
     |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
```

**Comment by vlopezj** [comment]
I checked using a `HasCallStack` constraint on `computeElimHeadType`, and the OP's error is indeed produced when doing anti-unification (https://github.com/agda/agda/commit/4717912820a7,  `master`)

```
 Checking book.14-propositional-truncation (/home/victor/phd/repos/agda/sandbox/2021-03-05_book_error/Agda/book/14-propositional-truncation.agda).
An internal error has occurred. Please report this as a bug.
Location of the error: __IMPOSSIBLE__, called at src/full/Agda/TypeChecking/Conversion.hs:401:20 in Agda-2.6.2-EYwicQBVxvv1nwjgCahl7w:Agda.TypeChecking.Conversion
  computeElimHeadType, called at src/full/Agda/TypeChecking/Conversion.hs:781:12 in Agda-2.6.2-EYwicQBVxvv1nwjgCahl7w:Agda.TypeChecking.Conversion
```

I tested with https://github.com/agda/agda/commit/1309bdb41cc6cc9066f93bd5cc51cf9dfec7c236 from `future-hetcon-no-compat` (which removes the `antiUnify` code) and I get a type error on another, later module:

```
 Checking book.14-propositional-truncation-solutions (/home/victor/phd/repos/agda/sandbox/2021-03-05_book_error/Agda/book/14-propositional-truncation-solutions.agda).
 Checking book.15-univalence (/home/victor/phd/repos/agda/sandbox/2021-03-05_book_error/Agda/book/15-univalence.agda).
/home/victor/phd/repos/agda/sandbox/2021-03-05_book_error/Agda/book/15-univalence.agda:480,7-481,27
Σ (sec (ap pr1)) (λ a → retr (ap pr1)) should be a function type,
but it isn't
when checking that X Y are valid arguments to a function of type
is-emb pr1
```

So it seems that #5234 would solve this issue also.

**Comment by jespercockx** [comment]
After some further investigation this was not at all because of anti-unification, but rather because of a call to `computeElimHeadType` from `Rewriting.hs`. I have a fix incoming...

**Comment by Saizan** [comment]
@jespercockx Also test against https://github.com/HoTT-Intro/Agda/commit/24f6861188f1405df7c3f673b5192a3d17016254

**Comment by jespercockx** [comment]
@Saizan I just ran the test with that version. First I was getting an error in `09-equivalences.agda` that there's an abstract definition that's missing a type signature (which is due to a recent change I guess). After adding the missing type signature to the function `is-equiv-map-right-absorption-prod`, I managed to check up to `15-univalence.agda`, where I'm getting the same error as the one mentioned by @vlopezj (the one that `Σ (sec (ap pr1)) (λ a → retr (ap pr1)) should be a function type`). So the internal error in `14-propositional-trunctation.agda` seems to be gone.
